### PR TITLE
Implements TEventClass methods and supporting functions.

### PR DESCRIPTION
**Pull request by OmniBlade**

### CODE REVIEW

**Review by ghost — DISMISSED** [review]

File: src/game/engine/gametypes.h [hunk 1]
```diff
@@ -324,20 +324,19 @@ DEFINE_ENUMERATION_OPERATORS(RadioMessageType);
...
-    ATTACH_2 = 2, // 2 ??
-    ATTACH_4 = 4, // 4 seems to be attatch to MapTriggers
-    ATTACH_8 = 8, // 8 seems to be attatch to HouseTriggers
-    ATTACH_16 = 16 // 16 seems to be attatch to LogicTriggers
+    ATTACH_1 = 1 << 0, // 1 ??
+    ATTACH_2 = 1 << 1, // 2 ??
+    ATTACH_4 = 1 << 2, // 4 seems to be attatch to MapTriggers
+    ATTACH_8 = 1 << 3, // 8 seems to be attatch to HouseTriggers
+    ATTACH_16 = 1 << 4 // 16 seems to be attatch to LogicTriggers
 };
```
Comment (tomsons26) [inline src/game/engine/gametypes.h:332]: These are ATTACH_CELL, ATTACH_OBJECT, ATTACH_MAP, ATTACH_HOUSE, ATTACH_LOGIC

File: src/game/engine/gametypes.h [hunk 2]
```diff
@@ -324,20 +324,19 @@ DEFINE_ENUMERATION_OPERATORS(RadioMessageType);
...
 
 DEFINE_ENUMERATION_BITWISE_OPERATORS(AttachType);
 
 //TODO: Will need looking into to really confirm, but credit to tomsons for the list.
 enum NeedType
 {
-    NEED_NONE = 0,
-    NEED_FIRST = 0,
+    NEED_NONE,
```
Comment (tomsons26) [inline src/game/engine/gametypes.h]: should be NEED_NOTHING.

File: src/game/engine/tevent.cpp [hunk 3]
```diff
@@ -0,0 +1,579 @@
...
+        case TEVENT_ANY:
+            return NEED_NONE;
+        case TEVENT_DESTROYED_ALL_UNITS:
+            return NEED_HOUSE;
+        case TEVENT_DESTROYED_ALL_BUILDINGS:
+            return NEED_HOUSE;
+        case TEVENT_DESTROYED_ALL:
+            return NEED_HOUSE;
+        case TEVENT_CREDIT_EXCEED:
+            return NEED_HOUSE;
```
Comment (tomsons26) [inline src/game/engine/tevent.cpp]: TEVENT_CREDIT_EXCEED need should be NEED_NUMBER.

File: src/game/engine/tevent.cpp [hunk 4]
```diff
@@ -0,0 +1,579 @@
...
+        case TEVENT_ATTACHED_BRIDGE_DESTROYED:
+            return NEED_NONE;
+        case TEVENT_BUILDING_EXISTS:
+            return NEED_BUILDING;
+        default:
+            break;
+    }
+
+    return NEED_HOUSE; // Seems to be the default value.
```
Comment (tomsons26) [inline src/game/engine/tevent.cpp]: default is NEED_NOTHING

File: src/game/engine/tevent.cpp [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+        case TEVENT_ZONE_ENTERED:
+            ret |= ATTACH_MAP;
+        default:
+            break;
+    }
+
+    switch (tevent) {
+        case TEVENT_THIEVED_BY:
+        case TEVENT_HOUSE_DISCOVERED:
+        case TEVENT_ATTACKED:
```
Comment (tomsons26) [inline src/game/engine/tevent.cpp]: Does not belong here.
After this is fixed this should be mergeable

**Review by tomsons26 — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**